File: src/main/java/com/flaghacker/sttt/bots/MMBot.java
package com.flaghacker.sttt.bots;

import com.flaghacker.sttt.common.*;

import java.util.ArrayList;
import java.util.List;

import static java.lang.Double.NEGATIVE_INFINITY;
import static java.lang.Double.POSITIVE_INFINITY;
import static java.lang.Math.max;

public class MMBot implements Bot
{
	private final int depth;

	public MMBot(int depth)
	{
		this.depth = depth;
	}

	@Override
	public Coord move(Board board, Timer timer)
	{
		if (board.nextPlayer() == Player.ENEMY)
			board = board.flip();

		return negaMax(board, depth+1, NEGATIVE_INFINITY, POSITIVE_INFINITY, 1).move;
	}

	private ValuedMove negaMax(Board board, int depth, double a, double b, int player)
	{
		if (depth == 0 || board.isDone())
			return new ValuedMove(board.getLastMove(), player * value(board));

		List<Board> children = children(board);

		double bestValue = NEGATIVE_INFINITY;
		Coord bestMove = null;

		for (Board child : children)
		{
			double value = -negaMax(child, depth - 1, -b, -a, -player).value;

			if (value > bestValue || bestMove == null)
			{
				bestValue = value;
				bestMove = child.getLastMove();
			}
			a = max(a, value);
			if (a >= b)
				break;
		}

		return new ValuedMove(bestMove, bestValue);
	}

	private List<Board> children(Board board)
	{
		List<Coord> moves = board.availableMoves();
		List<Board> children = new ArrayList<>(moves.size());

		for (Coord move : moves)
		{
			Board child = board.copy();
			child.play(move);
			children.add(child);
		}

		return children;
	}

	private static class ValuedMove
	{
		public final Coord move;
		public final double value;

		public ValuedMove(Coord move, double value)
		{
			this.move = move;
			this.value = value;
		}
	}

	private double value(Board board)
	{
		if (board.isDone())
			return Double.POSITIVE_INFINITY * playerSign(board.wonBy());

		double value = 0;

		for (Coord coord : Coord.list())
			value += TILE_VALUE * tileFactor(coord.os()) * tileFactor(coord.om()) * playerSign(board.tile(coord));

		for (int om = 0; om < 9; om++)
			value += MACRO_VALUE * tileFactor(om) * playerSign(board.macro(om));

		return value;
	}

	private static final double TILE_VALUE = 1;
	private static final double MACRO_VALUE = 10e9;

	private static final double CENTER_FACTOR = 4;
	private static final double CORNER_FACTOR = 3;
	private static final double EDGE_FACTOR = 1;

	private int playerSign(Player player)
	{
		switch (player)
		{
			case NEUTRAL:
				return 0;
			case PLAYER:
				return 1;
			case ENEMY:
				return -1;
		}

		throw new IllegalStateException();
	}

	private double tileFactor(int o)
	{
		int x = o % 3;
		int y = o / 3;

		if (x == 1 && y == 1)
			return CENTER_FACTOR;
		if (x == 1 || y == 1)
			return EDGE_FACTOR;
		return CORNER_FACTOR;
	}

	@Override
	public String toString()
	{
		return "MMBot";
	}
}

File: src/main/kotlin/com/flaghacker/sttt/bots/MMBot.kt
package com.flaghacker.sttt.bots

import com.flaghacker.sttt.common.*

import java.util.ArrayList

import java.lang.Double.NEGATIVE_INFINITY
import java.lang.Double.POSITIVE_INFINITY
import java.lang.Math.max

class MMBot(private val depth: Int) : Bot {
    override fun move(board: Board, timer: Timer): Byte? {
        return negaMax(board, depth + 1, NEGATIVE_INFINITY, POSITIVE_INFINITY, playerSign(board.nextPlayer())).move
    }

    private class ValuedMove(val move: Byte, val value: Double)

    private fun negaMax(board: Board, depth: Int, a: Double, b: Double, player: Int): ValuedMove {
        if (depth == 0 || board.isDone())
            return ValuedMove(board.lastMove()!!, player * value(board))

        val children = children(board)

        var bestValue = NEGATIVE_INFINITY
        var bestMove: Byte? = null

        var newA = a
        for (child in children) {
            val value = -negaMax(child, depth - 1, -b, -newA, -player).value

            if (value > bestValue || bestMove == null) {
                bestValue = value
                bestMove = child.lastMove()
            }
            newA = max(newA, value)
            if (newA >= b)
                break
        }

        return ValuedMove(bestMove!!, bestValue)
    }

    private fun children(board: Board): List<Board> {
        val moves = board.availableMoves()
        val children = ArrayList<Board>(moves.size)

        for (move in moves) {
            val child = board.copy()
            child.play(move)
            children.add(child)
        }

        return children
    }

    private fun value(board: Board): Double {
        if (board.isDone())
            return java.lang.Double.POSITIVE_INFINITY * playerSign(board.wonBy())

        return (0..80).sumByDouble {
            TILE_VALUE * tileFactor(it % 9) * tileFactor(it / 9) * playerSign(board.tile(it)).toDouble()
        } + (0..8).sumByDouble {
            MACRO_VALUE * tileFactor(it) * playerSign(board.macro(it)).toDouble()
        }
    }

    private fun playerSign(player: Player) = when (player) {
        Player.NEUTRAL -> 0
        Player.PLAYER -> 1
        Player.ENEMY -> -1
    }

    private fun tileFactor(o: Int): Double {
        val x = o % 3
        val y = o / 3

        if (x == 1 && y == 1)
            return CENTER_FACTOR
        return if (x == 1 || y == 1) EDGE_FACTOR else CORNER_FACTOR
    }

    override fun toString() = "MMBotJava"

    companion object {

        private val TILE_VALUE = 1.0
        private val MACRO_VALUE = 10e9

        private val CENTER_FACTOR = 4.0
        private val CORNER_FACTOR = 3.0
        private val EDGE_FACTOR = 1.0
    }
}
